### PR TITLE
Zanzana: List then search implementation

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What is this feature?**

This PR brings back search with list implementation after the schema change.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/services/authz/zanzana/client/client.go [hunk 1]
```diff
@@ -142,16 +142,16 @@ func newItemChecker(res *authzextv1.ListResponse) authz.ItemChecker {
...
 	return c.authzext.List(ctx, &authzextv1.ListRequest{
 		Subject:   id.GetUID(),
-		Group:     req.Group,
-		Verb:      utils.VerbList,
-		Resource:  req.Resource,
-		Namespace: req.Namespace,
+		Group:     req.GetGroup(),
+		Verb:      req.GetVerb(),
+		Resource:  req.GetResource(),
+		Namespace: req.GetNamespace(),
```
Comment (kalleep) [inline pkg/services/authz/zanzana/client/client.go:154]: We now pass in authzextv1.ListRequest so we don't have to create a new one and move over properties. We just have to set subject for the one we pass in here

File: pkg/services/dashboards/service/zanzana.go [hunk 2]
```diff
@@ -195,3 +206,97 @@ func (dr *DashboardServiceImpl) checkDashboardsBatch(ctx context.Context, query
 
 	return result, nil
 }
+
+// findDashboardsZanzanaList implements "List, then search" strategy. It first retrieve a list of resources
+// with given type available to the user and then passes that list as a filter to the search query.
+func (dr *DashboardServiceImpl) findDashboardsZanzanaList(ctx context.Context, query dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
```
Comment (kalleep) [inline pkg/services/dashboards/service/zanzana.go:212]: I don't think this one behaves correctly. When query type is not set we need to search both dashboards and folders. This is not happening here

Comment (alexanderzobnin) [reply pkg/services/dashboards/service/zanzana.go:212]: Yes, this is true, I need to fix it.

File: pkg/services/dashboards/service/zanzana.go [hunk 3]
```diff
@@ -85,13 +90,19 @@ func (dr *DashboardServiceImpl) findDashboardsZanzanaCompare(ctx context.Context
 	return first.result, first.err
 }
 
+type checkDashboardsFn func(context.Context, dashboards.FindPersistedDashboardsQuery, []dashboards.DashboardSearchProjection, int64) ([]dashboards.DashboardSearchProjection, error)
+
 func (dr *DashboardServiceImpl) findDashboardsZanzana(ctx context.Context, query dashboards.FindPersistedDashboardsQuery) ([]dashboards.DashboardSearchProjection, error) {
-	return dr.findDashboardsZanzanaCheck(ctx, query)
+	if len(query.Title) <= listQueryLengthThreshold || query.Limit > listQueryLimitThreshold {
+		return dr.findDashboardsZanzanaCheck(ctx, query, dr.checkDashboardsCompile)
```
Comment (kalleep) [inline pkg/services/dashboards/service/zanzana.go:97]: So for this one we only have to call compile for each resource once, right now we call it every time we have new results right?

Comment (kalleep) [reply pkg/services/dashboards/service/zanzana.go:97]: What I am trying to say is that we should be able to construct the checkers once and re-use them when we need to re fetch more data from the db

File: pkg/services/dashboards/service/zanzana.go [hunk 4]
```diff
@@ -207,43 +208,53 @@ func (dr *DashboardServiceImpl) checkDashboardsBatch(ctx context.Context, query
 	return result, nil
 }
 
-func (dr *DashboardServiceImpl) checkDashboardsCompile(ctx context.Context, query dashboards.FindPersistedDashboardsQuery, searchRes []dashboards.DashboardSearchProjection, remains int64) ([]dashboards.DashboardSearchProjection, error) {
+func (dr *DashboardServiceImpl) getCheckCompileFn(ctx context.Context, query dashboards.FindPersistedDashboardsQuery) checkDashboardsFn {
```
Comment (kalleep) [inline pkg/services/dashboards/service/zanzana.go:211]: we should return `(checkDashboardsFn, error)` so we don't have to call search if anything goes wrong here

**Review by kalleep — APPROVED** [review]
LGTM, left one more thing but otherwise we are good to go

**Pull request merged**